opencog/atoms/distvalue/DistributionalValue.cc

```diff
@@ -0,0 +1,455 @@
+/*
+ * opencog/truthvalue/DistributionalValue.cc
+ *
+ * Copyright (C) 2018 SingularityNet
+ * All Rights Reserved
+ *
+ * Written by Roman Treutlein
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License v3 as
+ * published by the Free Software Foundation and including the exceptions
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <float.h>
+#include <math.h>
+#include <stdio.h>
+
+#include <opencog/util/numeric.h>
+
+#include <opencog/atoms/distvalue/DistributionalValue.h>
+#include <opencog/atomspace/AtomSpace.h>
+
+#include <boost/range/combine.hpp>
+#include <iomanip>
+
+using namespace opencog;
+
+count_t DistributionalValue::DEFAULT_K = 800.0;
+
+DistributionalValue::DistributionalValue()
+	: Value(DISTRIBUTIONAL_VALUE)
+{}
+
+DistributionalValue::DistributionalValue(const DVCounter &dvctr)
+	: Value(DISTRIBUTIONAL_VALUE)
+{
+	for (auto elem : dvctr)
+	{
+		if (elem.second <= 0)
+			throw std::invalid_argument("DV cant have an element with count 0.");
+	}
+	_value = dvctr;
+}
+
+//Create a DV from the Parameters of a SimpleTV
+//Not recommended as it results in a DV with only 1 singleton set
+//which in some calculations is unusalbe as the chance of overlaps in the
+//Sets/Intervals is small
+DistributionalValue::DistributionalValue(double mode,double conf)
+	: Value(DISTRIBUTIONAL_VALUE)
+{
+	confidence_t cf = std::min(conf, 0.9999998);
+	double count = (DEFAULT_K * cf / (1.0 - cf));
+	//DV Interval with count 0 is undefined
+	count = std::max(count, 0.0000002);
+	DVKey k{Interval{mode}};
+	_value[k] = count;
+}
+
+DistributionalValuePtr DistributionalValue::createDV(const DVCounter &dvctr)
+{
+	return std::make_shared<const DistributionalValue>(dvctr);
+}
+
+DistributionalValuePtr DistributionalValue::createDV(double mode,
+                                                     double conf)
+{
+	return std::make_shared<const DistributionalValue>(mode,conf);
+}
+
+DistributionalValuePtr
+DistributionalValue::UniformDistributionalValue(const DVKey &k,int c)
+{
+	DVCounter dvctr;
+	dvctr[k] = c;
```

Comment:
This seems incorrect. This will cause a vector to be created, consisting of exactly k+1 bins.  The width of each bin will be exactly zero.  The location of each bin will be at 0.0.  Surely this is neither intended nor desirable.